### PR TITLE
[Java] CWE-601: Add Spring URL Redirect ResponseEntity sink

### PR DESCRIPTION
**Pull request by p0wn4j**
Add Open URL Redirect attack sinks by the `ResponseEntity` class in the Spring Framework.

### CODE REVIEW

**Comment by smowton** [comment]
Are you applying to the bounty program with this one?

**Comment by p0wn4j** [comment]
Yes I do.

**Comment by smowton** [comment]
OK, will hold off on reviewing til the security lab have weighed in

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp:25]: Change this to `the <code>bad</code> methods`

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:90]:
```suggestion
      argument = cie.getArgument([0, 1]) and
```

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:97]: Why are you asking for the `headers` or `location` calls to be a direct child of something of type `SpringResponseEntity`? When is a `headers` or `location` argument not a sink?

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:124]: Give this a more informative name, `springUrlRedirectTaintStep`

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp:24]:
```suggestion
The <code>bad</code> methods show an HTTP request parameter being used directly 
```

**Comment by smowton** [comment]
Test looks to be failing

**Comment by p0wn4j** [comment]
Locally running, test are not failing. Can I see what tests are failing?

**Comment by atorralba** [comment]
> Locally running, test are not failing. Can I see what tests are failing?

The failing test is `ql/java/ql/test/experimental/query-tests/security/CWE-601/SpringUrlRedirect.qlref`.

Seems that a bunch of results are missing:

```
-| SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String |
-| SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String | SpringUrlRedirect.java:91:27:91:49 | create(...) |
-| SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String |
-| SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String | SpringUrlRedirect.java:108:68:108:78 | httpHeaders |
-| SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String | SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String |
-| SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:91:27:91:49 | create(...) | semmle.label | create(...) |
-| SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String | semmle.label | httpHeaders [post update] [<map.value>, <element>] : String |
-| SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:108:68:108:78 | httpHeaders | semmle.label | httpHeaders |
-| SpringUrlRedirect.java:91:27:91:49 | create(...) | SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | SpringUrlRedirect.java:91:27:91:49 | create(...) | Potentially untrusted URL redirection due to $@. | SpringUrlRedirect.java:89:38:89:55 | redirectUrl | user-provided value |
-| SpringUrlRedirect.java:108:68:108:78 | httpHeaders | SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | SpringUrlRedirect.java:108:68:108:78 | httpHeaders | Potentially untrusted URL redirection due to $@. | SpringUrlRedirect.java:104:39:104:56 | redirectUrl | user-provided value |
```

Maybe you need to rebase/merge `main` to see the failures. You also have some conflicts that need resolution.

**Comment by smowton** [comment]
`java/ql/test/experimental/query-tests/security/CWE-601/SpringUrlRedirect.qlref` still failing. Rebase on latest main?

**Comment by p0wn4j** [comment]
After merge conflict fixing my tests ran ok. Now I  again rebase on lastest main.

**Comment by smowton** [comment]
Reproduced the problem: I was running codeql 2.5.5 and was getting the same results as you; having upgraded to 2.5.7 I now see the failure

**Comment by smowton** [comment]
It was down to a recently-fixed extractor bug that used to misname static inner classes like `Map$Entry` as `Map<>$Entry`. In this case `RequestEntity` no longer needed the superfluous `<>`. I've pushed a commit that fixes the issue.

**Comment by p0wn4j** [comment]
Ok, thank you very much :)

**Review by smowton — APPROVED** [review]

**Pull request merged**
